src/test/test_main.cpp

```diff
@@ -0,0 +1,122 @@
+        PASS(logic_callback_add_pre_draw_once),
+        PASS(logic_callback_add_post_draw_once),
+        PASS(logic_callback_single_frame),
+        PASS(logic_callback_forever),
+        PASS(logic_callback_remove),
+        PASS(logic_callback_update),
+
+        /*
+        PASS(test_slow_test_false),
+        PASS(test_slow_test_false),
+        PASS(test_slow_test_false),
+        PASS(test_slow_test_false),
+        FAIL(test_slow_test_false),
+        FAIL(test_slow_test_false),
+        FAIL(test_slow_test_false),
+        PASS(test_slow_test_true),
+        PASS(test_slow_test_true),
+        PASS(test_slow_test_true),
+        PASS(test_slow_test_true),
+        FAIL(test_slow_test_true),
+        FAIL(test_slow_test_true),
+        FAIL(test_slow_test_true),
+        FAIL(test_slow_test_true),
+        */
+    };
+    int size = sizeof(tests) / sizeof(UnitTest);
+    int passed = 0;  // counts only tests that should pass
+    int failed = 0;  // counts only tests that should fail
+    int passed_expected = 0;
+    int failed_expected = 0;
+    int skipped = 0;
+
+    // std::random_device rd;
+    // int seed = rd();
+    // printf("Shuffling with seed %d", seed);
+    // std::mt19937 generator(seed);
+    // std::shuffle(std::begin(tests), std::end(tests), generator);
+
+    auto start = std::chrono::steady_clock::now();
+    int current = 0;
+    for (UnitTest test: tests) {
+        current++;
+        if (test.skip) {
+            skipped++;
+            printf(CLEAR "\r" YELLOW "  skipping " RESET "%s\n", test.name);
+            continue;
+        }
+        printf(CLEAR "\r  %d/%d: " YELLOW " testing " RESET "%s\r", current, size, test.name);
+        fflush(stdout);
+
+        Result res = test.test();
+        if (res == FAIL_EXT) {
+            skipped++;
+            printf("  %s fails before actual test\n", test.name);
+            continue;
+        }
+        if (test.failing) {
+            failed_expected++;
+            if (res == FAIL) {
+                failed++;
+            } else {
+                printf("  %s passed even though it shouldn't\n", test.name);
+            }
+        } else {
+            passed_expected++;
+            if (res == PASS) {
+                passed++;
+            } else {
+                printf("  %s failed even though it shouldn't\n", test.name);
+            }
+        }
+    }
+    auto end = std::chrono::steady_clock::now();
+    std::chrono::nanoseconds elapsed = end - start;
+
+    printf(CLEAR);
+    printf(GREEN  "passed:" RESET "               %3d\n", passed);
+    printf(RED    "unexpected failures:" RESET "  %3d\n", (passed_expected - passed));
+    printf(RED    "unexpected successes:" RESET " %3d\n", (failed_expected - failed));
+    printf(YELLOW "expected failures:" RESET "    %3d\n", failed);
+    printf(YELLOW "skipped:" RESET "              %3d\n", skipped);
+    printf("%d tests in %.2f ms\n", size - skipped, (elapsed.count() / 1e6));
```

Comment:
chrono, I personally wouldn't use it, but that doesn't mean its bad. This isn't critical code so I'm fine with this. It can get messy when you want to change to different time-formats, but you kinda skipp it here. But it is the "C++ way", and the usage looks good.